### PR TITLE
Improve S1694: Remove part about protected constructor

### LINKED ISSUE #8795: Improve S1694: Remove part about protected constructor
S1694 asks to do two things:

Change abstract class with only abstract members to an interface.
Change abstract class without abstract members to a class with protected constructor. We should do the exact opposite on this point.

## Why?

If a class should be inherited by design, it should be marked as such in the class declaration. It's an explicit contract on the class level.

The rule asks to create a random member (constructor) placed somewhere inside the class (hard to find by-definition), with a specific visibility and hope that users will infer what the contract was supposed to be from that. 

That is not practical from code usability and maintainability point of view.

Attempt to instantiate abstract class provides clear explanation, that is actionable (find implementing class, or create one)
> CS0144	Cannot create an instance of the abstract type or interface 'Sample1'

While attempt to instantiate a class with protected constructor provides generic message that is same for private and protected constructors. Making it unclear if the class just should not be created.
> CS0122	'Sample2.Sample2()' is inaccessible due to its protection level

## What?

Update S1694 to limit its scope to `abstract class -> interface` part and remove the `abstract class -> protected constructor part`.

Make the rule SonarWay.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #8795

Should be merged with RSPEC: https://github.com/SonarSource/rspec/pull/3960

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeAbstract.cs [hunk 1]
```diff
@@ -1,113 +1,107 @@
 ﻿using System;
 using System.Collections.Generic;
 
-namespace Tests.Diagnostics
+public abstract partial class A
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/ClassShouldNotBeAbstract.cs:4]: This is a really interesting test in the Noncompliant version:
```cs
public abstract partial class A // Noncompliant
{
    public abstract void X();
}
public abstract partial class A // Noncompliant
{
    public abstract void Y();
}
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeAbstract.cs [hunk 2]
```diff
@@ -18,113 +18,79 @@
...
-        private static bool HasInheritedAbstractMembers(INamedTypeSymbol symbol)
-        {
-            var baseTypes = symbol.BaseType.GetSelfAndBaseTypes().ToList();
-            var abstractMethods = baseTypes.SelectMany(GetAllAbstractMethods);
-            var baseTypesAndSelf = baseTypes.Concat(new[] { symbol }).ToList();
-            var overrideMethods = baseTypesAndSelf.SelectMany(GetAllOverrideMethods);
-            var overriddenMethods = overrideMethods.Select(m => m.OverriddenMethod);
-            var stillAbstractMethods = abstractMethods.Except(overriddenMethods);
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ClassShouldNotBeAbstract : SonarDiagnosticAnalyzer<SyntaxKind>
```
Comment (CristianAmbrosini) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeAbstract.cs:24]: The name of this class is not the best IMO, what about `AbstractClassShouldContainConcreteMethods`, or a shorter version of it

**Review by CristianAmbrosini — CHANGES_REQUESTED** [review]
A few lines can be removed without failing the UTs

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeAbstract.cs [hunk 3]
```diff
@@ -18,113 +18,79 @@
...
-                if (node is ClassDeclarationSyntax classDeclaration)
-                {
-                    context.ReportIssue(Rule, classDeclaration.Identifier.GetLocation(), "class", message);
-                }
-
-                if (RecordDeclarationSyntaxWrapper.IsInstance(node))
+                var symbol = (INamedTypeSymbol)c.Symbol;
+                if (symbol.IsClass()
+                    && symbol.IsAbstract
+                    && !HasInheritedAbstractMembers(symbol)
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeAbstract.cs:42]: symbol.IsAbstract is a performance shortcut

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ClassShouldNotBeAbstract.cs:42]: `HasInheritedAbstractMembers` is a good catch! It's already excluded by `symbol.BaseType.Is(KnownType.System_Object)`  check

**Review by CristianAmbrosini — APPROVED** [review]
LGTM!

**Pull request merged**